CodegenProposal.md

```diff
@@ -0,0 +1,1474 @@
+```swift
+switch size.value {
+case .SMALL: break
+case .AVERAGE: break
+case .LARGE: break
+default: break 
+// or
+case .none: break
+}
+```
+See [GraphQLEnum.swift](Sources/ApolloAPI/GraphQLEnum.swift) for implementation details.
+
+# `InputObject` Generation
+
+Input objects will be generated for each `input` type in the schema that is used in an argument for any of the operations defined in your application. Input objects are structs that are backed by a `InputDict` struct that stores the values for the fields on the input object in a dictionary. This allows for `InputObject`s to be treated as values types but use copy-on-write semantics under the hood.
+
+Nullable fields on input objects are represented using `GraphQLNullable` to allow for both `null` and `nil` values. 
+
+Following the [Input Coercion rules](https://spec.graphql.org/draft/#sec-Input-Objects.Input-Coercion) from the GraphQL spec, the server defined default value for a field will be used when passing `nil`.  Nullable fields on input objects are represented using `GraphQLNullable` to allow for both `null` and `nil` values.  For non-nullable fields, if the schema provides a default value, the field will be represented as an optional to allow for `nil` to be passed. 
+
+**Examples:**
+
+Nullable field with no default value:
+```graphql
+input MyInput {
+  size: RelativeSize
+}
+```
+```swift
+struct MyInput: InputObject {
+  public private(set) var dict: InputDict
+
+  init(size: GraphQLNullable<RelativeSize> = nil) {
+    dict = InputDict(["size": size])
+  }
+
+  var size: GraphQLNullable<RelativeSize> {
+    get { dict["size"] }
+    set { dict["size"] = newValue }
+  }
+}
+```
+Nullable field with a default value:
+```graphql
+input MyInput {
+  size: RelativeSize = SMALL
+}
+```
+```swift
+struct MyInput: InputObject {
+  public private(set) var dict: InputDict
+
+  init(size: GraphQLNullable<RelativeSize>) { ... }
+
+  /// If `.none`, defaults to server-provided default value (.SMALL)
+  var size: GraphQLNullable<RelativeSize> { ... }
+}
+```
+Non-nullable field with no default value:
+```graphql
+input MyInput {
+  size: RelativeSize!
+}
+```
+```swift
+struct MyInput: InputObject {
+  public private(set) var dict: InputDict
+
+  init(size: RelativeSize) { ... }
+  
+  var size: RelativeSize { ... }
+}
+```
+Non-nullable field with a default value:
+```graphql
+input MyInput {
+  size: RelativeSize! = SMALL
+}
+```
+```swift
+struct MyInput: InputObject {
+  public private(set) var dict: InputDict
+
+  init(size: RelativeSize?) { ... }
+  
+  /// If `nil`, defaults to server-provided default value (.SMALL)
+  var size: RelativeSize? { ... }
+}
+```
+> Note that we are not generating these fields with the provided default values. This is to account for default values that may change on the schema in the future. See [Generate Default Parameter Values for `InputObject` Default Values](#generate-default-parameter-values-for-inputobject-default-values) for more discussion.
```

Comment:
One other thing worth considering: What happens if there's a default value provided in the schema when it's generated, but then the default value is removed on the server later?

Reply:
That would be a breaking change. Since backwards compatibility is enforced as a standard, breaking changes are "supposed" to be avoided and you would make a new field and deprecate the old one if you needed to do this. 

If developers choose to make breaking changes, the Apollo client library behavior is undefined, and we shouldn't be expected to handle those situations gracefully.

In this case, the server would return a graphQL error response.

Reply:
Probably worth noting that explicitly in this proposal